### PR TITLE
Support for `type:platform`

### PR DESCRIPTION
**Pull request by anurse**
Adds support for using `type:platform` in Publish to distinguish which package should be burned in as the shared framework.

---
This change is [Reviewable](https://reviewable.io/reviews/dotnet/cli/1932)

### CODE REVIEW

File: src/Microsoft.DotNet.ProjectModel/Graph/LibraryDependencyType.cs [hunk 1]
```diff
@@ -5,45 +5,24 @@
...
-        public static LibraryDependencyType Default = LibraryDependencyType.Parse("default");
+        public static LibraryDependencyType Empty = new LibraryDependencyType();
+        public static LibraryDependencyType Default = new LibraryDependencyType("default");
+        public static LibraryDependencyType Build = new LibraryDependencyType("build");
+        public static LibraryDependencyType Platform = new LibraryDependencyType("platform");
 
-        public static LibraryDependencyType Build = LibraryDependencyType.Parse("build");
+        public string Value { get; }
 
-        public LibraryDependencyTypeFlag Flags { get; private set; }
```
Comment (anurse) [inline src/Microsoft.DotNet.ProjectModel/Graph/LibraryDependencyType.cs]: We weren't using most of the flags, so @davidfowl and I discussed it and decided to just use the keywords instead.

File: src/Microsoft.DotNet.Compiler.Common/Executable.cs [hunk 2]
```diff
@@ -133,15 +130,30 @@ private void WriteRuntimeConfig(LibraryExporter exporter)
...
+                    .ToList();
+                if(redistPackage.Count > 0)
                 {
-                    var framework = new JObject(
-                        new JProperty("name", redistExport.Library.Identity.Name),
-                        new JProperty("version", redistExport.Library.Identity.Version.ToNormalizedString()));
-                    runtimeOptions.Add("framework", framework);
+                    if(redistPackage.Count > 1)
+                    {
+                        throw new InvalidOperationException("Multiple packages with type: \"platform\" were specified!");
```
Comment (pakrym) [inline src/Microsoft.DotNet.Compiler.Common/Executable.cs]: I feel this is not right place for this check. Maybe LibraryManager so VS can show it as error in project.json?

File: src/Microsoft.DotNet.ProjectModel/Graph/LibraryDependencyType.cs [hunk 3]
```diff
@@ -5,45 +5,24 @@
...
         }
 
-        public bool HasFlag(LibraryDependencyTypeFlag flag)
-        {
-            return (Flags & flag) != 0;
-        }
+        public static LibraryDependencyType Parse(string value) => new LibraryDependencyType(value.ToLowerInvariant());
+        public override int GetHashCode() => Value.GetHashCode();
+        public override bool Equals(object obj) => obj is LibraryDependencyType && Equals((LibraryDependencyType)obj);
+        public bool Equals(string other) => string.Equals(Value, other, StringComparison.Ordinal);
```
Comment (pakrym) [inline src/Microsoft.DotNet.ProjectModel/Graph/LibraryDependencyType.cs]: Why do you hate newlines? :)

File: src/Microsoft.DotNet.ProjectModel/Compilation/LibraryAsset.cs [hunk 4]
```diff
@@ -17,6 +17,7 @@ public struct LibraryAsset
         public string ResolvedPath { get; }
         public string FileName => Path.GetFileName(RelativePath);
         public Action<Stream, Stream> Transform { get; set; }
+        public bool IsPlaceholder => FileName.Equals("_._");
```
Comment (pakrym) [inline src/Microsoft.DotNet.ProjectModel/Compilation/LibraryAsset.cs]: Should we filter them out in LibraryExporter?

Comment (anurse) [inline src/Microsoft.DotNet.ProjectModel/Compilation/LibraryAsset.cs]: Actually, I think we're going to need them in the exporter. We need to call out the difference between no assets were **compatible** and no assets are **needed** and preserve that in the `deps` file.

**Comment by anurse** [comment]
:up: :date:

/cc @pakrym @davidfowl

File: Microsoft.DotNet.Cli.sln [hunk 5]
```diff
@@ -59,10 +57,6 @@ Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.DotNet.Cli.Build.
 EndProject
 Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "dotnet-compile.UnitTests", "test\dotnet-compile.UnitTests\dotnet-compile.UnitTests.xproj", "{920B71D8-62DA-4F5E-8A26-926C113F1D97}"
 EndProject
-Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "TestAppWithArgs", "TestAssets\TestProjects\TestAppWithArgs\TestAppWithArgs.xproj", "{DA8E0E9E-A6D6-4583-864C-8F40465E3A48}"
```
Comment (anurse) [inline Microsoft.DotNet.Cli.sln]: I got tired of only some test projects being in the solution, so I just removed the only two that were. #easymode

File: Microsoft.DotNet.Cli.sln [hunk 6]
```diff
@@ -35,8 +35,6 @@ Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "EndToEnd", "test\EndToEnd\E
 EndProject
 Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.DotNet.Tools.Test.Utilities", "test\Microsoft.DotNet.Tools.Tests.Utilities\Microsoft.DotNet.Tools.Test.Utilities.xproj", "{E4F46EAB-B5A5-4E60-9B9D-40A1FADBF45C}"
 EndProject
-Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "TestProjects", "TestProjects", "{713CBFBB-5392-438D-B766-A9A585EF1BB8}"
```
Comment (pakrym) [inline Microsoft.DotNet.Cli.sln]: unremove

File: NuGet.Config [hunk 7]
```diff
@@ -9,8 +9,4 @@
   <activePackageSource>
     <add key="AspNetCIDev" value="https://www.myget.org/F/aspnetcidev/api/v3/index.json" />
   </activePackageSource>
-  <packageRestore>
-    <add key="enabled" value="False" />
-    <add key="automatic" value="False" />
-  </packageRestore>
```
Comment (anurse) [inline NuGet.Config]: Not technically required for this PR but now that we have some tooling, it kinda works. I'll probably revert before merging.

File: src/Microsoft.DotNet.Compiler.Common/BindingRedirectGenerator.cs [hunk 8]
```diff
@@ -109,7 +109,11 @@ private static XElement GetOrAddElement(XContainer parent, XName elementName)
 
         private static AssemblyRedirect[] CollectRedirects(IEnumerable<LibraryExport> dependencies)
         {
-            var allRuntimeAssemblies = dependencies.SelectMany(d => d.RuntimeAssemblies).Select(GetAssemblyInfo).ToArray();
+            var allRuntimeAssemblies = dependencies
+                .SelectMany(d => d.RuntimeAssemblyGroups.GetGroup(string.Empty))
```
Comment (pakrym) [inline src/Microsoft.DotNet.Compiler.Common/BindingRedirectGenerator.cs]: I don't like passing string.empty around, it doesn't mean a lot. Maybe do optional parameter?

Comment (pakrym) [inline src/Microsoft.DotNet.Compiler.Common/BindingRedirectGenerator.cs]: Or GetDefaultGroup?

File: src/Microsoft.DotNet.ProjectModel/DependencyContextBuilder.cs [hunk 9]
```diff
@@ -145,11 +139,17 @@ private static CompilationOptions GetCompilationOptions(CommonCompilerOptions co
                     export.Library.Hash,
                     assemblies,
                     libraryDependencies,
-                    serviceable
-                   );
+                    serviceable);
             }
         }
 
+        private RuntimeAssetGroup CreateRuntimeAssetGroup(LibraryAssetGroup arg)
```
Comment (pakrym) [inline src/Microsoft.DotNet.ProjectModel/DependencyContextBuilder.cs]: arg  :)

File: src/Microsoft.DotNet.ProjectModel/Resolution/LibraryManager.cs [hunk 10]
```diff
@@ -128,6 +134,19 @@ public IList<DiagnosticMessage> GetAllDiagnostics()
...
             }
 
+            if (platformLibraries.Count > 1)
+            {
+                foreach (var platformLibrary in platformLibraries)
+                {
+                    AddDiagnostics(
+                        messages,
+                        platformLibrary,
+                        "Multiple dependencies are marked with type 'platform'. Only one dependency can have this type.",
```
Comment (pakrym) [inline src/Microsoft.DotNet.ProjectModel/Resolution/LibraryManager.cs]: We got the list already, maybe we should dump it into error?

File: src/Microsoft.Extensions.DependencyModel/DependencyContextWriter.cs [hunk 11]
```diff
@@ -203,13 +203,11 @@ private JObject WriteTargetLibrary(Library library)
...
                 AddDependencies(libraryObject, runtimeLibrary.Dependencies);
-                AddRuntimeAssemblies(libraryObject, runtimeLibrary.Assemblies);
-                AddResourceAssemblies(libraryObject, runtimeLibrary.ResourceAssemblies);
 
-                if (runtimeLibrary.NativeLibraries.Any())
-                {
-                    libraryObject.Add(DependencyContextStrings.NativeLibrariesKey, WriteAssetList(runtimeLibrary.NativeLibraries));
-                }
+                // Add runtime-agnostic assets
+                AddAssets(libraryObject, DependencyContextStrings.RuntimeAssembliesKey, runtimeLibrary.RuntimeAssemblyGroups.FirstOrDefault(r => string.IsNullOrEmpty(r.Runtime)));
```
Comment (pakrym) [inline src/Microsoft.Extensions.DependencyModel/DependencyContextWriter.cs]: Make a method out of `RuntimeAssemblyGroups.FirstOrDefault(r => string.IsNullOrEmpty(r.Runtime)`

File: src/Microsoft.Extensions.DependencyModel/RuntimeAsset.cs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+        public static RuntimeAsset Create(string path)
+        {
+            var name = System.IO.Path.GetFileNameWithoutExtension(path);
+            if (name == null)
+            {
+                throw new ArgumentException($"Provided path has empty file name '{path}'", nameof(path));
+            }
+
+            if (name.EndsWith(NativeImageSufix))
```
Comment (pakrym) [inline src/Microsoft.Extensions.DependencyModel/RuntimeAsset.cs]: I don't like this logic being run for native libraries, supposedly they would never end with `.ni` but I feel this check doesn't belong here.

File: src/Microsoft.DotNet.ProjectModel/DependencyContextBuilder.cs (hunk 9 above)
Comment (anurse) [inline src/Microsoft.DotNet.ProjectModel/DependencyContextBuilder.cs]: yaaaaaaaarg

image

File: src/Microsoft.Extensions.DependencyModel/RuntimeAsset.cs (hunk 12 above)
Comment (anurse) [inline src/Microsoft.Extensions.DependencyModel/RuntimeAsset.cs]: We could just have `Name` have the file name alone, with file extension intact.

I really want to keep the common type here between Runtime and Native assets.

File: src/Microsoft.Extensions.DependencyModel/DependencyContext.cs [hunk 13]
```diff
@@ -107,5 +116,27 @@ public int GetHashCode(T obj)
                 return obj.Name.GetHashCode();
             }
         }
+
+        private IEnumerable<RuntimeAsset> ResolveAssets(string runtimeIdentifier, Func<RuntimeLibrary, IEnumerable<RuntimeAssetGroup>> groupSelector)
+        {
+            var fallbacks = RuntimeGraph.FirstOrDefault(f => f.Runtime == runtimeIdentifier);
+            var rids = Enumerable.Concat(new[] { runtimeIdentifier }, fallbacks?.Fallbacks ?? Enumerable.Empty<string>());
```
Comment (anurse) [inline src/Microsoft.Extensions.DependencyModel/DependencyContext.cs]: ?? ? : ? ::: ?? ?? ::: new [] ?? { } ??

**Comment by anurse** [comment]
@dotnet-bot test CentOS7.1 x64 Debug Build please

**Comment by anurse** [comment]
@dotnet-bot test this please

**Comment by piotrpMSFT** [comment]
@dotnet-bot test this please.

**Comment by anurse** [comment]
@dotnet-bot test RHEL7.2 x64 Release Build please
@dotnet-bot test Windows_2016 x64 Release Build please

**Comment by anurse** [comment]
Merging without 2016. It's SO SLOW

**Pull request merged**